### ISSUE TITLE
Add a justification message input to http get and http post LogiX nodes

### ISSUE THREAD

**Issue by Earthmark** [Neos-Metaverse/NeosPublic#2878]
### Is your feature request related to a problem? Please describe.

Currently the "proper" way to do network events requires asking first, then doing a network event if access is granted. However it seems much more common that people send and just kind of pray that it works.

This is made worse with that the network ask prompt doesn't use the same url filtering rules as the http* nodes, so you need to do some odd string filtering as well. Currently the request node does not filter port, route, or protocol, and providing those can sometimes make the user request and http request misalign and the user is asked twice.
This is likely a bug, but I would like to request a different kind of fix instead.

Along with users sometimes blocking the url after the first of a few requests, your need to have a justification node next to each web request, even if it's not the first one.

Websockets however, currently elegantly get around this issue.
Websockets have a justification field built in, so you populate that once and wherever a request isn't allowed it automatically asks. (The websocket state machine has it's own issues, but in terms of asking it's quite nice!)

### Describe the solution you'd like

Add an input pin to the http* nodes for request justification. If the url isn't authorized that justification is shown to the user through the normal means.

The request url could no longer have sync issues, and this would remove the complexity from adding a justification check. As this would make it one fixed string value to say why, I think this may help reduce the number of context less web requests that occur (although the url formatting issue is likely also a cause of that).

### Describe alternatives you've considered

Just fixing the request url formatting bug, this is doable but I would like providing a connection reason to be as easy as possible. I feel making it easy is how you make it more common.

### Additional context

I'll add picture context about the request urls in a bit here, I'm on my phone and kept forgetting to post this issue/request.

**Comment by ProbablePrime** [comment]
What is wrong with the Request Host Access Node here? This is meant to be the way you ask for permission

> each web request
This will probably just cause double the problems the current solution has.

> Websockets however, currently elegantly get around this issue. Websockets have a justification field built in
Can you elaborate on where this is? Websockets shouldn't be used to get around using Request Host Access.

> Currently the request node does not filter port, route, or protocol, and providing those can sometimes make the user request and http request misalign and the user is asked twice.
Fixing this(i was taking a poke the other day). Seems to be a better use of time. We can change up the filtering and even re-make the consent dialog to look and read better instead.

**Comment by Earthmark** [comment]
When I pulse the impulse on the left I get this message. 
![image](https://user-images.githubusercontent.com/1251609/131201085-bf54206e-2399-4304-8323-6bda65972eaa.png)
Of note: when granted the Is Host Access Allowed does not auto update.

If I ping the node on the right, I get this prompt no matter what.
![image](https://user-images.githubusercontent.com/1251609/131201120-310177fa-7028-40c7-8ec9-8a8758074b8e.png)

**Comment by Earthmark** [comment]
The websocket client has an AccessReason built in.
![image](https://user-images.githubusercontent.com/1251609/131201173-c77abca2-2ce5-43e8-8d00-a5a7c6dd71ac.png)

**Comment by ProbablePrime** [comment]
Ahh thank you for the pictures. 

From this it looks like the following list should be looked at:
1. Double checking Is Host Access Allowed works / Checking if it needs an updating Relay
2. Better Documentation on how the Request Host Access Node is meant to be used
3. Better filtering for Ports/Subdomains on the host access node

The access reason on the websocket client makes sense as this is a component as it needs to maintain an active connection. The HTTP nodes don't they close the connection when done.

I'm sorry here but I think we're trying to add features here instead of:
1. Fixing Bugs
2. Tweaking Things that already exist

and therefore I'm not certain this is a good idea.

**Comment by Earthmark** [comment]
> What is wrong with the Request Host Access Node here? This is meant to be the way you ask for permission

I personally feel it should be easiest to do the right thing, if every request can ask for access, then every request should have the ability for you to fill out the reason as to why you're requesting access. Right now the system explicitly requires there to be a case where the user can be asked for access without a reason, and that feels kinda strange to me.

> This will probably just cause double the problems the current solution has.

To my knowedge this is already the case, just you don't get a reason as to why. Either way you can still block and block forever, I'm just asking for an easy way to get a message into the prompt.
 
> Can you elaborate on where this is? Websockets shouldn't be used to get around using Request Host Access.

Picture sent.

> Fixing this(i was taking a poke the other day). Seems to be a better use of time. We can change up the filtering and even re-make the consent dialog to look and read better instead.

Sweet! That would be nice, I would like to try to remove the no-justification edge case still but making it easier to use the justification node would be nice still!

**Comment by ProbablePrime** [comment]
> I personally feel it should be easiest to do the right thing, if every request can ask for access, then every request should have the ability for you to fill out the reason as to why you're requesting access. Right now the system explicitly requires there to be a case where the user can be asked for access without a reason, and that feels kinda strange to me.

I feel this is just bad Logix. You should always use request host access/is host access allowed before starting any HTTP operations. 

I dont really want to provide the option for users to be forced to explain every request they're doing. 

Right now, accessing one host can be requested for a good reason: "To syncronize your data with potato.com"

Whereas adding reasons to each request would turn into:
* Potato.com would like to get your data
* Potato.com would like to update your data
* Potato.com would like to check the daily price of potatos.

HTTP systems generally make a ton of requests and displaying consent for each and every one of them would just lead to a whole host of rpoblems.

**Comment by Earthmark** [comment]
> I feel this is just bad Logix. You should always use request host access/is host access allowed before starting any HTTP operations.

> I dont really want to provide the option for users to be forced to explain every request they're doing.

I feel these two are saying opposing things, the first says every http node should have a request in front of it, and the second is saying you don't want users to need to put a request node in front of every http node... I think we are miscommunicating on something.

> Right now, accessing one host can be requested for a good reason: "To syncronize your data with potato.com"

I think I see where the confusion is, I'm not asking for a change to the way authorization occurs, I'm not asking for it to prompt the user on every request. I'm just asking for a way to get a message into the prompt if the user is prompted during a request. No other changes in behavior are being requested here.

I expect the same message will get routed into every request.

**Comment by ProbablePrime** [comment]
I see, In this case I stand by what I'm saying but I'll explain this better.

In my opinion, you should use the Request Host Access/IsHostAccessAllowed **once** at the start of your gadget's **entire** lifecycle. For example if its a UI/Thing similar to lets say the mentor signal. You might check for host access when it is initially spawned only. 
* If it is allowed, you don't need to worry again about access. Don't bother with any additional host access nodes
* If it isn't allowed, display a message on your UI stating this with a button that starts the request host access flow

This way you don't have to plumb a justification string into all of your nodes.

**Comment by Earthmark** [comment]
I think if I could rely on that first case I would accept that, I feel that as users can choose to reject a request at any point by canceling the host authorization, I would need to have a request access node in front of every http node anyways.

If the request access node gets fixed parsing wise that'll at least resolve part of the issue, I can deal with the extra nodes around my http requests.

**Comment by Frooxius** [comment]
I definitely agree that this would lead to bad LogiX and worsen the problem. A single application can have potentially lots of HTTP request nodes and instrumenting each one of those with a reason seems to be against the Don't Repeat Yourself principle.

The intended use is that you first ensure you have the permission from given user, before running rest of everything. Once you have the permission you activate the rest of your code and you can run all the requests you want. You shouldn't need to add it in front of each node, I'd recommend structuring your code around that, rather than defensively checking of every single request.

This is analogous to the Websocket - the reason is asked for establishing connection. Every Websocket message you then send on that connection then just goes through. What you're asking is analogous to adding the reason to every Websocket message request. The main thing that's different here is that HTTP doesn't have this initial connection establishment. 

Instead we'd focus on improving the other problems that you mention instead. We might probably expand on the reason for connection as well to be a bit more formal - add more information, explanation, maybe logo and so on. This is part of the problem too - imagine you had to update every single of your HTTP nodes with this information and keep those in sync. If we add more nodes there as well it would require us to update all the different nodes too with everything.

**Comment by ProbablePrime** [comment]
In the case of suddenly revoking the access, you should perhaps be able to FireOnFalse the IsHostAccessAllowed node. 

We should have a separate issue for better host filtering.

As for adding a reason to each node, let's see what Froox says.

**Comment by Earthmark** [comment]
Perhaps this is a reason to have a central http client component you can make requests through? I've been wondering about cookies in the http clients as well, as currently I don't think they're directly supported but I think the websocket client does support them. Either way that's outside of the scope of the feature request.

The one thing I don't want to do is prompt the user with a justificationless web request dialog, I'll ponder based on this if I can make a safe way to do that based on what's described here, I still think there's edge cases but maybe there's a way around them. Perhaps if I could make a http request node fail instead of asking blankly that'd be a way around my concern of asking with a blank dialog.

**Comment by Frooxius** [comment]
I see what you mean with the random message without justification. 

Central component for HTTP might be interesting idea perhaps.  Another alternative is that the nodes will simply ignore any requests if permission isn't given first (we'd convert old nodes to preserve old behavior to avoid breaking existing setups, but any new setup would have to use this new approach).

**Comment by ProbablePrime** [comment]
On cookies:
Cookies are not often used for API requests or what we usually call Service to Service requests and are instead used for Web Browser Requests that return HTML. This isn't usually what our HTTP nodes are used for.

With collections its likely we'll get full header support with the HTTP nodes which would allow you to persist tokens, cookies, parameters etc to headers etc.

I'm particularly fond of OAuth Tokens and JWTs. 

This is of course a better discussion for another issue though.

On a HTTP Component, I'm not sure components are the way forward for everything. 

I do like Froox's conversion Idea. 

Regardless here, I'll make some additional tutorials.

**Comment by Earthmark** [comment]
Cookies arn't supported in the http nodes right now at all to my knowedge, I was doing random testing and found that they were with websockets but that was a bit ago. Given we don't have a way to do authorization or pass auth headers, having a cookie added would be a good way to keep track of a single users requests in a way that's harder to skim, you'd need to specifically use the client on the remote users machine instead of only needing to steal a token inside the world data model.

I do like the migration approach, I don't personally like that it will make the logix more complicated, but it would reinforce the expected pattern and wouldn't prompt the user unexpectedly.

This may also support the issue... I can't find the issue now, but it was about artificially making an access dialog and asking if the user is OK with something. They could craft an artificial url and provide a reason, then watch if that artificial url is approved or not and move on, never connecting to any urls related to the non-existent remote resource.

**Comment by Earthmark** [comment]
On second thought it doesn't make the logix more complicated, it just makes the problematic pattern not viable, the 'correct' logix would be identical.

**Comment by ProbablePrime** [comment]
Sorry, I didn't mean to say that they were supported.

Both Cookies and Headers will require collections.

Generically, we should avoid cookies. They're not useful for the patterns Neos provides in terms of HTTP. Headers are much better.

**Comment by ProbablePrime** [comment]
With the new changes to URIs and host filtering I'll be closing this one. 

Also see #2881 which should also help.
